File: pset6/credit/credit.py
from cs50 import *
from enum import Enum

class CreditCard(Enum):
    AmericanExpress = 'AMEX'
    MasterCard = 'MASTERCARD'
    Visa = 'VISA'
    Undefined = ''

CARD_LEN = {
    CreditCard.AmericanExpress: [15],
    CreditCard.MasterCard: [16],
    CreditCard.Visa: [13, 16]
}

CARD_PREFIX = {
    CreditCard.AmericanExpress: ["34", "37"],
    CreditCard.MasterCard: ["51", "52", "53", "54", "55"],
    CreditCard.Visa: ["4"]
}

def main():
    card_number = get_int("Number: ")
    card_number_str = str(card_number)
    card_type = get_card_type(card_number_str)
    if card_type != CreditCard.Undefined:
        print(card_type.value)
    else:
        print("INVALID")

def get_checksum(number_str):
    error_val = -1
    result = 0
    # Check type
    if not isinstance(number_str, str):
        return error_val
    # Check string is a positive number string
    try:
        test_num = int(number_str)
        if test_num < 0:
            return error_val
    except:
        return error_val
    # Calculate checksum
    for i, c in enumerate(reversed(number_str)):
        if (i % 2) == 0:
            num = int(c)
        else:
            num = int(c) * 2
        result += (num % 10)
        if num >= 10:
            result += (num // 10)
    return result

def get_card_type(card_number_str):
    # Check type
    if not isinstance(card_number_str, str):
        return CreditCard.Undefined
    # Check card number length and prefix
    result = CreditCard.Undefined
    card_len = len(card_number_str)
    for card_type in CreditCard:
        if not (card_type in CARD_LEN):
            continue
        if not (card_len in CARD_LEN[card_type]):
            continue
        found = False
        if not (card_type in CARD_PREFIX):
            continue
        for card_prefix in CARD_PREFIX[card_type]:
            if card_number_str.startswith(card_prefix):
                found = True
                break
        if found:
            result = card_type
            break
    # If length and prefix of card number is invalid, skip checksum test
    if result == CreditCard.Undefined:
        return CreditCard.Undefined
    # Check the checksum of card number
    checksum = get_checksum(card_number_str)
    if (checksum % 10) == 0:
        return result
    else:
        return CreditCard.Undefined

if __name__ == "__main__":
    main()

File: pset1/credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>

int main(void)
{
    long card_num;
    do
    {
        card_num = get_long("Number: ");
    }
    while (card_num < 0);

    // get card number length
    int card_len = 0;
    do
    {
        card_len++;
    } while (card_num > pow(10, card_len));
    if (card_len != 13 && card_len != 15 && card_len != 16) {
        printf("INVALID\n");
        return -1;
    }

    // get checksum of card number
    int checksum = 0;
    for (int i = 0; i < card_len; i++) {
        long lower_mutipler = pow(10, i);
        long digit = (card_num / lower_mutipler) % 10;
        if (i % 2 == 1) {
            checksum = checksum + (digit * 2) / 10 + (digit * 2) % 10;
        } else {
            checksum = checksum + digit;
        }
    }
    if (checksum % 10 != 0) {
        printf("INVALID\n");
        return -1;
    }

    // get prefix 2 digit of car number
    int start_two_digit = card_num / pow(10, card_len - 2);
    if ((start_two_digit == 34 || start_two_digit == 37) && card_len == 15) {
        printf("AMEX\n");
    } else if ((51 <= start_two_digit && start_two_digit <= 55) && card_len == 16) {
        printf("MASTERCARD\n");
    } else if (start_two_digit / 10 == 4 && (card_len == 13 || card_len == 16)) {
        printf("VISA\n");
    } else {
        printf("INVALID\n");
    }
}
